libraries/apollo-compiler/src/main/kotlin/com/apollographql/apollo3/compiler/codegen/kotlin/helpers/DataClass.kt

```diff
@@ -62,17 +65,193 @@ fun TypeSpec.Builder.makeDataClassFromProperties(properties: List<PropertySpec>)
+    if (propertySpecs.isNotEmpty()) {
+      withToStringImplementation()
+    }
+  }
+
+  if (generateMethods.contains(COPY)) {
+    if (propertySpecs.isNotEmpty()) {
+      withCopyImplementation()
+    }
+  }
+}
+
+fun TypeSpec.Builder.withCopyImplementation(): TypeSpec.Builder {
+  val type = build()
+  val constructorParamNames = type.primaryConstructor?.parameters?.map { it.name }?.toSet() ?: return this
+  val typeName = type.name ?: return this
+  val className = ClassName("", typeName)
+  val constructorProperties = propertySpecs
+      .excludeInternalProperties()
+      .filter { constructorParamNames.contains(it.name) }
+
+  check(constructorProperties.size == constructorParamNames.size) {
+    "Cannot generate copy method for class with constructor arguments that are not properties"
+  }
+
+  return addFunction(FunSpec.builder(Identifier.copy)
+      .apply {
+        constructorProperties.forEach {
+          addParameter(ParameterSpec.builder(it.name, it.type).defaultValue("this.%L", it.name).build())
+        }
+      }
+      .returns(className)
+      .addCode(CodeBlock.builder()
+          .add("return %T(", className)
+          .apply {
+            constructorProperties.forEach {
+              add("%L,", it.name)
+            }
+          }
+          .add(")")
+          .build()
+      )
+      .build())
+}
+
+fun TypeSpec.Builder.withEqualsImplementation(): TypeSpec.Builder {
+  fun equalsCode(property: PropertySpec): CodeBlock {
+    return CodeBlock
+        .builder()
+        .addStatement("this.%L == other.%L", property.name, property.name).build()
+  }
+
+  fun methodCode(typeName: String?): CodeBlock {
+    if (typeName == null) {
+      return CodeBlock.builder()
+          .addStatement("return·other != null·&&·other::class·==·this::class")
+          .build()
+    }
+    val className = ClassName("", typeName)
+    return CodeBlock.builder()
+        .beginControlFlow("if (other == this)")
+        .addStatement("return true")
+        .endControlFlow()
+        .beginControlFlow("if (other is %T)", className)
+        .apply {
+          if (propertySpecs.isEmpty()) {
+            add("return true")
+          } else {
+            add("return %L", propertySpecs
+                .excludeInternalProperties()
+                .map { equalsCode(it) }
+                .joinToCode("&& ")
+            )
+          }
+        }
+        .endControlFlow()
+        .addStatement("return false")
+        .build()
+  }
+
+  return addFunction(FunSpec.builder(Identifier.equals)
+      .addModifiers(KModifier.OVERRIDE)
+      .addParameter("other", KotlinSymbols.Any.copy(nullable = true))
+      .returns(KotlinSymbols.Boolean)
+      .addCode(methodCode(build().name))
+      .build())
+}
+
+fun TypeSpec.Builder.withHashCodeImplementation(): TypeSpec.Builder = apply {
+  fun hashPropertyCode(property: PropertySpec) =
+      CodeBlock.builder()
+          .addStatement("${Identifier.__h} *= 31")
```

Comment:
This matches the Kotlin implementation: https://github.com/JetBrains/kotlin/blob/master/compiler/ir/ir.tree/src/org/jetbrains/kotlin/ir/util/DataClassMembersGenerator.kt#L235